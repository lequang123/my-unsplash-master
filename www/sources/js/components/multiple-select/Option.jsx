import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class Option extends Component {
    static propTypes = {
        itemData: PropTypes.object.isRequired,
        id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
        onChange: PropTypes.func
    };

    constructor(props) {
        super(props);

        this.state = {
            isExpandedSubData: true
        };
    }

    onChange = event => {
        event.stopPropagation();

        const { itemData, onChange } = this.props;

        const newItemData = {
            key: itemData.key,
            value: itemData.value,
            checked: !itemData.checked
        };
        onChange && onChange(newItemData);
    };

    onClickExpandIcon = () => {
        this.setState(prevState => ({
            isExpandedSubData: !prevState.isExpandedSubData
        }));
    };

    renderSubData = subDataList => {
        return (
            <ul className='sub-data'>
                {
                    this.state.isExpandedSubData &&
                    subDataList.map((item, key) => <li key={`sub-${key}`}>{item}</li>)
                }
            </ul>
        );
    };

    render() {
        const { itemData, id } = this.props;
        const isHasSubData = itemData.subData && itemData.subData.length > 0;
        return (
            <li className="multiple-select-item">
                <input
                    id={id}
                    type="checkbox"
                    className="option-checkbox"
                    checked={itemData.checked}
                    onChange={this.onChange}
                />
                <label className="option-label" htmlFor={id}>
                    {itemData.value}
                </label>
                {
                    isHasSubData &&
                    <span className={`expand-icon fa ${this.state.isExpandedSubData ? 'fa-caret-down' : 'fa-caret-right'}`} onClick={this.onClickExpandIcon} />
                }
                {
                    isHasSubData &&
                    this.renderSubData(itemData.subData)
                }
            </li>
        );
    }
}