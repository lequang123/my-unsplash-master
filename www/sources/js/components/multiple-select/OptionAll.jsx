import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Option from './Option';
import { DEFAULT_CHECKED} from './constants';

export default class OptionAll extends Component {
    static propTypes = {
        onChange: PropTypes.func,
        id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
        checked: PropTypes.bool,
        textOptions: PropTypes.object
    };

    static defaultProps = {
        checked: DEFAULT_CHECKED
    };

    constructor(props) {
        super(props);

        this.itemData = {
            key: 'All',
            value: this.props.textOptions['Texts']
        };
    }

    changeHandler = ({ checked }) => {
        this.props.onChange && this.props.onChange(checked);
    };

    render() {
        const { id, checked } = this.props;
        this.itemData.checked = checked;

        return (
            <ul className="multiple-select-options multiple-select-option-all">
                <Option
                    id={`${id}-optionItemAll`}
                    itemData={this.itemData}
                    onChange={this.changeHandler}
                />
            </ul>
        );
    }
}