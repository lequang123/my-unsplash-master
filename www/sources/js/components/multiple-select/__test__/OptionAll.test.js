import React from 'react';
import OptionAll from '../OptionAll';
import Option from '../Option';
import { shallow } from 'enzyme';

describe('OptionAll', () => {
    it('should have correct snapshots', () => {
        // Arrange
        const props = {
            textOptions: {
                Texts: 'Texts'
            }
        };

        // Action
        const optionAll = shallow(<OptionAll {...props} />);

        // Assert
        expect(optionAll).toMatchSnapshot();
    });

    it('changeHandler should run correctly', () => {
        // Arrange
        const checked = true;
        const props = {
            textOptions: {
                Texts: 'Texts'
            },
            onChange: jest.fn()
        };

        const optionAll = shallow(<OptionAll {...props} />);

        // Action
        optionAll.instance().changeHandler({ checked });

        // Assert
        expect(props.onChange).toHaveBeenCalledWith(checked);
    });

    it('render should return correct components', () => {
        // Arrange
        const props = {
            id: 'id',
            checked: true,
            textOptions: {
                Texts: 'Texts'
            }
        };

        const changeHandler = jest.fn();
        const optionAll = shallow(<OptionAll {...props} />);
        optionAll.instance().changeHandler = changeHandler;
        
        // Action
        var result = shallow(optionAll.instance().render());

        // Assert
        expect(result.props().className).toBe('multiple-select-options multiple-select-option-all');
        expect(result.find(Option).props().id).toBe('id-optionItemAll');
        expect(result.find(Option).props().itemData).toEqual({
            key: 'All',
            value: props.textOptions.Texts,
            checked: props.checked
        });
        expect(result.find(Option).props().onChange).toEqual(changeHandler);
    });
});
