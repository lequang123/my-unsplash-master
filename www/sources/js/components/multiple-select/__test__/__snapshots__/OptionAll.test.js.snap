// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionAll should have correct snapshots 1`] = `ShallowWrapper {}`;
